sentry-graphql/src/main/java/io/sentry/graphql/ExceptionReporter.java

```diff
@@ -0,0 +1,146 @@
+import graphql.language.AstPrinter;
+import graphql.schema.DataFetchingEnvironment;
+import io.sentry.Hint;
+import io.sentry.IHub;
+import io.sentry.SentryEvent;
+import io.sentry.SentryLevel;
+import io.sentry.exception.ExceptionMechanismException;
+import io.sentry.protocol.Mechanism;
+import io.sentry.protocol.Request;
+import io.sentry.protocol.Response;
+import java.util.HashMap;
+import java.util.Map;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class ExceptionReporter {
+  private final boolean isSpring;
+
+  public ExceptionReporter(final boolean isSpring) {
+    this.isSpring = isSpring;
+  }
+
+  private static final @NotNull String MECHANISM_TYPE = "GraphqlInstrumentation";
+
+  public void captureThrowable(
+      final @NotNull Throwable throwable,
+      final @NotNull ExceptionDetails exceptionDetails,
+      final @Nullable ExecutionResult result) {
+    final @NotNull IHub hub = exceptionDetails.getHub();
+    final Mechanism mechanism = new Mechanism();
+    mechanism.setType(MECHANISM_TYPE);
+    mechanism.setHandled(false);
+    final Throwable mechanismException =
+        new ExceptionMechanismException(mechanism, throwable, Thread.currentThread());
+    final SentryEvent event = new SentryEvent(mechanismException);
+    event.setLevel(SentryLevel.FATAL);
+
+    final Hint hint = new Hint();
+    setRequestDetailsOnEvent(hub, exceptionDetails, event);
+
+    if (result != null) {
+      @NotNull Response response = new Response();
+      Map<String, Object> responseBody = result.toSpecification();
+      response.setData(responseBody);
+      event.getContexts().setResponse(response);
+    }
+
+    hub.captureEvent(event, hint);
+  }
+
+  private void setRequestDetailsOnEvent(
+      final @NotNull IHub hub,
+      final @NotNull ExceptionDetails exceptionDetails,
+      final @NotNull SentryEvent event) {
+    hub.configureScope(
+        (scope) -> {
+          final @Nullable Request scopeRequest = scope.getRequest();
+          if (scopeRequest != null) {
+            setDetailsOnRequest(hub, exceptionDetails, scopeRequest);
+            event.setRequest(scopeRequest);
+          } else {
+            Request newRequest = new Request();
+            setDetailsOnRequest(hub, exceptionDetails, newRequest);
+            event.setRequest(newRequest);
+          }
+        });
+  }
+
+  private void setDetailsOnRequest(
+      final @NotNull IHub hub,
+      final @NotNull ExceptionDetails exceptionDetails,
+      final @NotNull Request request) {
+    request.setApiTarget("graphql");
+
+    if (exceptionDetails.isSubscription() || !isSpring) {
+      final @NotNull Map<String, Object> data = new HashMap<>();
+
+      data.put("data", exceptionDetails.getQuery());
+
+      if (hub.getOptions().isSendDefaultPii()) {
+        data.put("variables", exceptionDetails.getVariables());
+      }
```

Comment:
Why is it guarded by `sendDefaultPii`? Relay will do PII.

Reply:
Because in this case it's easy to do in the SDK. Should we not do it if we get a chance? We check for the flag in los of other places.

Reply:
The `variables` keys are used to match the response `data` to do PII stripping.
So if the variables contain `a: b`, and the response protocol contains `a`, it strips out.
If you don't send the keys at all, Relay won't be able to do so.

Reply:
> This feature is opt-in for this reason as well

At the moment this is auto configured for spring and the act of opting in is adding a dependency which also changes when people start using SAGP for their backend to auto install dependencies. Do we make adding request / response opt-in?

For Spring (regardless of GraphQL) `sendDefaultPii` opts you into sending the request payload. So do we simply move the whole request / response attaching code for GraphQL inside an if guarded by `sendDefaultPii` as well?

Reply:
@adinauer instead of guarding against `sendDefaultPii` I'd rather introduce a new flag (maybe?).
On Mobile it's called `captureFailedRequests` because the Apollo lib is specifically for GraphQL errors anyway.
If errors are already captured for GraphQL errors already and we are just enriching them with more context, maybe we need to control that only, with an extra flag/configuration or whatever makes sense, but the request and response payload should be opt-in.

Reply:
The existing integration for graphql-java (before this PR) already reports (some of the) exceptions to Sentry. Now guarding this behind a flag seems odd. I'd rather make this consistent with how we do it in Spring and guard by `sendDefaultPii` as well. Especially since spring-graphql is our prime target and this way subscriptions would behave the same as queries and mutations.

Reply:
@adinauer capturing errors wouldn't change, it's about adding the body (request and response) or not, this part should be opt-in.